src/experiments/settings/api.js

```diff
@@ -0,0 +1,176 @@
+      return value;
+    }
+    let type = Services.prefs.getPrefType(name);
+    switch (type) {
+      case Services.prefs.PREF_STRING:
+        return Services.prefs.getCharPref(name, value);
+      case Services.prefs.PREF_INT:
+        return Services.prefs.getIntPref(name, value);
+      case Services.prefs.PREF_BOOL:
+        return Services.prefs.getBoolPref(name, value);
+      default:
+        throw new Error("Unknown type");
+    }
+  }
+}
+const SETTING_TYPE = "setting_config";
+const SETTING_PREFIX = "settingRolloutConfig_";
+const settingManager = {
+  async init() {
+    await ExtensionSettingsStore.initialize();
+  },
+  async add(id, settingConfig) {
+    await this.init();
+    await ExtensionSettingsStore.addSetting(id, SETTING_TYPE, `${SETTING_PREFIX}${settingConfig.name}`, settingConfig);
+  },
+  async getSetting(settingName) {
+    await this.init();
+    return await ExtensionSettingsStore.getSetting(SETTING_TYPE, `${SETTING_PREFIX}${settingName}`);
+  },
+  async getSettingConfig(settingName) {
+    const setting = await this.getSetting(settingName);
+    return setting.value;
+  },
+  async get(settingName) {
+    const settingConfig = await this.getSettingConfig(settingName);
+    const statePref = await prefManager.getUserPref(settingConfig.statePref);
+    let currentState = null;
+    Object.keys(settingConfig.states).forEach((stateKey) => {
+      const state = settingConfig.states[stateKey];
+      if (state.id === statePref) {
+        currentState = stateKey;
+      }
+    });
+    return currentState;
+  },
+  async set(id, settingName, value, includePrefs = true) {
+    const config = await this.getSettingConfig(settingName);
+    let match = {};
+    let state;
+    // If we match a state set the state pref and prefs
+    // The state pref can't be cleared
+    if (value in config.states) {
+      state = config.states[value];
+      // If we aren't clearing the prefs we want these too
+      if (includePrefs) {
+        match = state.prefs;
+      }
+      match = Object.assign(match, state.persistPrefs, {
+        [config.statePref]: state.id
+      });
+    }
+
+    // set prefs to state or reset
+    let prefs = {};
+    for (let pref of Object.keys(config.prefTypes)) {
+      prefManager.setPref(pref, match[pref] || undefined, config.prefTypes[pref]);
+    }
+
+    return true;
+  },
+  // Clear all non persistant prefs and remove the addon.
+  async clear(id, finalState = "uninstalled") {
+    await this.init();
+    const types = await ExtensionSettingsStore.getAllForExtension(id, SETTING_TYPE);
+    for (let key of types) {
+      const settingName = key.replace(new RegExp(`^${SETTING_PREFIX}`), "");
+      const state = await this.get(settingName);
+      const config = await this.getSettingConfig(settingName);
+      // If we don't have a valid state or it's not a minus state then set uninstalled
+      if (!config.states[state] || config.states[state].id >= 0) {
```

Comment:
So currently that means once the user clicks "OK" and gets into state `-4`, uninstalling or disabling in the add-on manager will not overwrite that.

Is that intended? It doesn't seem like it.

This is again the dilemma of "we're uninstalling but did the user really uninstall?". I'm not sure how to solve this :(

Reply:
This is the intended, as soon as the user clicks OK then we consider that as a final state. The only change we make is clearing the non persistent prefs.

Reply:
As I was putting it in though I felt the same, it's very hard to track what users are intending to do here. `-4` is basically just a different uninstall giving us slightly more data to reason with on the next study.

Reply:
You probably want to note that briefly in the _comment for the "disabled" state.